metldata/artifacts_rest/artifact_dao.py

```diff
@@ -78,6 +78,29 @@ def _get_dao_name(*, artifact_name: str, class_name: str) -> str:
 
         return f"art_{artifact_name}_class_{class_name}"
 
+    async def get_all_resource_tags(self) -> set[str]:
+        """Retrieve resource tags for all artifacts currently present in the db.
+
+        A resource tag combines artifact_name, class_name and resource into a string
+        separated by #, i.e. it has the form 'artifact_name#class_name#resource'.
+        This is done to flatten the nested artifact information to simplify keeping track
+        of changes.
+        """
+
+        all_resource_tags = []
+        for artifact_type, class_name_dao in self._artifact_daos.items():
+            for class_name, resource_dao in class_name_dao.items():
+                # empty mapping should yield all resources
+                resource_ids = {
+                    resource.id_ async for resource in resource_dao.find_all(mapping={})
+                }
+                for resource_id in resource_ids:
+                    all_resource_tags.append(
+                        f"{artifact_type}#{class_name}#{resource_id}"
+                    )
+
+        return set(all_resource_tags)
```

Comment:
```suggestion
        all_resource_tags: set[str] = set()
        for artifact_type, class_name_dao in self._artifact_daos.items():
            for class_name, resource_dao in class_name_dao.items():
                # empty mapping should yield all resources
                all_resource_tags.update(
                    [
                        f"{artifact_type}#{class_name}#{resource.id_}"
                        async for resource in resource_dao.find_all(mapping={})
                    ]
                )

        return all_resource_tags
```
Just an idea, not necessary: you could make `all_resource_tags` a set from the start and update it with a list comprehension for each item in `class_name_dao`.

Reply:
From experience updating a set in a loop will get quite slow if it gets large enough.
Not knowing how big the information here might get, I decided to go with a list and just have the overhead once in the end.

Reply:
Why use sets at all? Shouldn't the resource ids be all unique anyway?

```python
        for artifact_type, class_name_dao in self._artifact_daos.items():
            for class_name, resource_dao in class_name_dao.items():
                # empty mapping should yield all resources
                resource_ids = [
                    resource.id_ async for resource in resource_dao.find_all(mapping={})
                ]
                all_resource_tags.extend(
                    f"{artifact_type}#{class_name}#{resource_id}"
                    for resource_id in resource_ids
                )
```

Also, why not use 3-tuples instead of string tags? Feels more natural and doesn't require a split operation later.

Reply:
Oh, totally forgot that I could just do it that way, makes thin easier.
Will get right to changing that